### ISSUE TITLE
error when load onnx model

### ISSUE THREAD

**Issue by fnhdx** [opencv/opencv#16475]
I am trying to load a onnx model which is converted from pytorch. 

`import cv2`
`model = cv2.dnn.readNetFromONNX('tmp.onnx')`

cv2.error: OpenCV(4.1.2) /io/opencv/modules/dnn/src/dnn.cpp:525: error: (-2:Unspecified error) Can't create layer "166" of type "Cast" in function 'getLayerInstance'

here is my model
[https://www.dropbox.com/s/sg0n1kdpbbio848/tmp.onnx?dl=0](url)

anyone have idea how to fix this problem? Thanks a lot.

**Comment by dkurt** [comment]
Are custom layers applicable in that case? See https://docs.opencv.org/master/dc/db1/tutorial_dnn_custom_layers.html

**Comment by fnhdx** [comment]
Is upsample layer in pytorch supported by opencv onnx?

**Comment by dkurt** [comment]
@fnhdx, According to error message, it says that "Cast" layer is not supported. "Upsample" should be parsed:

https://github.com/opencv/opencv/blob/3d14dd4e3953aa7b7e2ab3a388ca7bf08fe99d99/modules/dnn/src/onnx/onnx_importer.cpp#L956-L974

**Comment by fnhdx** [comment]
@dkurt It's strange. There is no layer called "Cast" in the pytorch model. Here is the model I used which is a standard unet model

`
class DoubleConv(nn.Module):

    def __init__(self, in_channels, out_channels):
        super().__init__()
        self.double_conv = nn.Sequential(
            nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1),
            nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True),
            nn.Conv2d(out_channels, out_channels, kernel_size=3, padding=1),
            nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True)
        )

    def forward(self, x):
        return self.double_conv(x)


class Down(nn.Module):

    def __init__(self, in_channels, out_channels):
        super().__init__()
        self.maxpool_conv = nn.Sequential(
            nn.MaxPool2d(2),
            DoubleConv(in_channels, out_channels)
        )

    def forward(self, x):
        return self.maxpool_conv(x)


class Up(nn.Module):

    def __init__(self, in_channels, out_channels, bilinear=True):
        super().__init__()

        # if bilinear, use the normal convolutions to reduce the number of channels
        if bilinear:
            # self.up = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)
            self.up = nn.Upsample(scale_factor=2, mode='nearest') 
        else:
            self.up = nn.ConvTranspose2d(in_channels // 2, in_channels // 2, kernel_size=2, stride=2)

        self.conv = DoubleConv(in_channels, out_channels)

    def forward(self, x1, x2):
        x1 = self.up(x1)
        # input is CHW
        diffY = torch.tensor([x2.size()[2] - x1.size()[2]])
        diffX = torch.tensor([x2.size()[3] - x1.size()[3]])

        x1 = F.pad(x1, [diffX // 2, diffX - diffX // 2,
                        diffY // 2, diffY - diffY // 2])
        x = torch.cat([x2, x1], dim=1)
        return self.conv(x)


class OutConv(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(OutConv, self).__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=1)

    def forward(self, x):
        return self.conv(x)

class UNet(nn.Module):
    def __init__(self, n_channels, n_classes, bilinear=True):
        super(UNet, self).__init__()
        self.n_channels = n_channels
        self.n_classes = n_classes
        self.bilinear = bilinear

        self.inc = DoubleConv(n_channels, 64)
        self.down1 = Down(64, 128)
        self.down2 = Down(128, 256)
        self.down3 = Down(256, 512)
        self.down4 = Down(512, 512)
        self.up1 = Up(1024, 256, bilinear)
        self.up2 = Up(512, 128, bilinear)
        self.up3 = Up(256, 64, bilinear)
        self.up4 = Up(128, 64, bilinear)
        self.outc = OutConv(64, n_classes)

    def forward(self, x):
        x1 = self.inc(x)
        x2 = self.down1(x1)
        x3 = self.down2(x2)
        x4 = self.down3(x3)
        x5 = self.down4(x4)
        x = self.up1(x5, x4)
        x = self.up2(x, x3)
        x = self.up3(x, x2)
        x = self.up4(x, x1)
        logits = self.outc(x)
        return logits

def get_UNet(config, **kwargs):

    model = UNet(n_channels=3, n_classes=6)


    return model

`

**Comment by ashishkrshrivastava** [comment]
@dkurt , seems like its because of 

> self.up = nn.ConvTranspose2d(in_channels` // `2, in_channels` // `2, kernel_size=2, stride=2)

and 

> x1 = F.pad(x1, [diffX` //` 2, diffX - diffX` // `2, diffY` // `2, diffY - diffY` //` 2]). 

Both `cast` and `floor` layer need to be implemented.

**Comment by dkurt** [comment]
@ashishkrshrivastava, probably, it's about shape values. Usually, they are constant and we just cast them to integers manually,
```
self.up = nn.ConvTranspose2d(int(in_channels//2), int(in_channels//2), kernel_size=2, stride=2)
```
So this is a workaround - export model but with `int(*)` casts.

However if it is impossible, it would be good feature to support. I believe this graph fusion for ONNX models can help to make it easier if we can determine which subgraph should be fused: https://github.com/opencv/opencv/pull/16291

**Comment by fnhdx** [comment]
@dkurt @ashishkrshrivastava, as your suggestions, I removed all the type conversions and simplified my model (see below). Unfortunately, I still got same error message 
cv2.error: OpenCV(4.1.2) /io/opencv/modules/dnn/src/dnn.cpp:525: error: (-2:Unspecified error) Can't create layer "131" of type "Cast" in function 'getLayerInstance'

`class DoubleConv(nn.Module):

    def __init__(self, in_channels, out_channels):
        super().__init__()
        self.double_conv = nn.Sequential(
            nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1),
            nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True),
            nn.Conv2d(out_channels, out_channels, kernel_size=3, padding=1),
            nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True)
        )

    def forward(self, x):
        return self.double_conv(x)


class Down(nn.Module):

    def __init__(self, in_channels, out_channels):
        super().__init__()
        self.maxpool_conv = nn.Sequential(
            nn.MaxPool2d(2),
            DoubleConv(in_channels, out_channels)
        )

    def forward(self, x):
        return self.maxpool_conv(x)

class Up(nn.Module):

    def __init__(self, in_channels, out_channels, bilinear=True):
        super().__init__()

        self.up = nn.Upsample(scale_factor=2, mode='nearest') 
        self.conv = DoubleConv(in_channels, out_channels)

    def forward(self, x1, x2):
        x1 = self.up(x1)
        x = torch.cat([x2, x1], dim=1)
        return self.conv(x)


class OutConv(nn.Module):

    def __init__(self, in_channels, out_channels):
        super(OutConv, self).__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=1)

    def forward(self, x):
        return self.conv(x)

class UNet(nn.Module):

    def __init__(self, n_channels, n_classes, bilinear=True):
        super(UNet, self).__init__()
        self.n_channels = n_channels
        self.n_classes = n_classes
        self.bilinear = bilinear

        self.inc = DoubleConv(n_channels, 64)
        self.down1 = Down(64, 128)
        self.down2 = Down(128, 256)
        self.down3 = Down(256, 256)
        self.up1 = Up(512, 128, bilinear)
        self.up2 = Up(256, 64, bilinear)
        self.up3 = Up(128, 64, bilinear)
        self.outc = OutConv(64, n_classes)


    def forward(self, x):
        x1 = self.inc(x)
        x2 = self.down1(x1)
        x3 = self.down2(x2)
        x4 = self.down3(x3)
        x = self.up1(x4, x3)
        x = self.up2(x, x2)
        x = self.up3(x, x1)
        logits = self.outc(x)
        return logits

def get_face_alignment_depthsplitblock(config, **kwargs):

    model = UNet(n_channels=3, n_classes=6)
    return model
`

**Comment by ashishkrshrivastava** [comment]
@fnhdx It would be really helpful if you upload your recent onnx model.

**Comment by fnhdx** [comment]
here is the model
https://www.dropbox.com/s/3yihghgh61vn8nj/tmp2.onnx?dl=0

**Comment by ashishkrshrivastava** [comment]
Hello @fnhdx ,My exported model with your code is little different from yours. can you please provide the code you used to export the model. 
Something like this
`torch.onnx.export(model, inputs, f, export_params=True, opset_version=11)`

**Comment by fnhdx** [comment]
@ashishkrshrivastava here is the code to export the model
`import torch.onnx

model_file = 'pathToModel/checkpoint_0.pth'
state_dict = torch.load(model_file, map_location='cpu')
model = state_dict['state_dict']
model = model.module

img = torch.randn(1, 3, 56, 56, requires_grad=True)
img = torch.tensor(img)
out = model(img)[0]
torch.onnx.export(model, img, 'tmp2.onnx', opset_version=11)`

**Comment by ashishkrshrivastava** [comment]
Thanks @fnhdx , can you please mention onnx version. 
```
import onnx
onnx.__version__
```

**Comment by fnhdx** [comment]
@ashishkrshrivastava 
I didn't actually install onnx in my pytorch environment. I am importing pytorch.onnx. Don't know what version they are using.

**Comment by dkurt** [comment]
Please test the changes from https://github.com/opencv/opencv/pull/16573 (now in 3.4 branch and will be ported to master this week.
```python
import numpy as np
import cv2 as cv

inp = np.random.standard_normal([1, 3, 56, 56]);
net = cv.dnn.readNet('tmp.onnx')
net.setInput(inp)
out = net.forward()
print(out.shape)
```
```
(1, 6, 56, 56)
```

**Comment by adityak2920** [comment]
Hey, 
I am also trying to load U-Net in OpenCV using onnx, but I am still getting this error.
``` Can't create layer "526" of type "Cast" in function 'getLayerInstance'```
OpenCV==4.2.0
torch==1.4.0

> @ashishkrshrivastava
> here is the model, thx
> https://www.dropbox.com/s/3yihghgh61vn8nj/tmp2.onnx?dl=0

I have also tried to load the same model, but it's throwing the same error.
Please guide me on what can I do?

**Comment by dkurt** [comment]
@adityak2920, you need to try master branch because the fix is merged after 4.2.0 release.

**Comment by adityak2920** [comment]
@ashishkrshrivastava @dkurt , Thanks for your answer. Now, I have installed OpenCV from master branch. But, now I am getting another error. 
```cv2.error: OpenCV(4.2.0-dev) /Users/adityakumar/opencv/modules/dnn/src/graph_simplifier.cpp:79: error: (-212:Parsing error) Input node with name 2885 not found in function 'getInputNodeId'```
Here is the link to my model: [model](https://drive.google.com/open?id=1bwCAM3HxWpl7OGNM_jhFwsHL4dfDiKnl)
Please, check what can i do to resolve it?

**Comment by ashishkrshrivastava** [comment]
@adityak2920 Can you please share the link or code of the model.
You can try [this](https://pytorch.org/hub/mateuszbuda_brain-segmentation-pytorch_unet/) model of unet, if that works for you.
Otherwise please create separate issue because the model you shared if a lot different from the model specified in this issue.

**Comment by adityak2920** [comment]
@ashishkrshrivastava Actually, I am trying to load FasterRCNN model not UNet. And I have created a new issue #16783 . Please review it and check where I am going wrong.

**Comment by istranic** [comment]
All, I'm having the same issue, and I can't quite identify the solution based on this thread. I exported a UNet model from Pytorch into ONNX using:

torch.onnx.export(...

I'm then loading the model using:

cv.dnn.readNetFromONNX(...

and I get the error:

Can't create layer "252" of type "Cast" in function 'cv::dnn::dnn4_v20191202::LayerData::getLayerInstance'

Any help is much appreciated.


VERSIONS:
OpenCv 4.2.0
Pytorch 1.4.0

**Comment by adityak2920** [comment]
@istranic you should install OpenCV using the master branch.

**Comment by istranic** [comment]
@adityak2920 that's 4.0.0?

**Comment by adityak2920** [comment]
@istranic no, that will be 4.2.0-dev

**Comment by istranic** [comment]
@adityak2920 Did you mean 4.3.0-dev?

I don't see 4.2.0-dev in https://docs.opencv.org/master/

**Comment by adityak2920** [comment]
I am talking about the master branch.

**Comment by adityak2920** [comment]
@istranic if you are on 4.2.0.34 then this error should not occur. You should also put opt_version=11 while exporting pytorch model to ONNX. If still, you are facing error then just install OpenCV from the master branch.

**Comment by istranic** [comment]
@adityak2920 I pip installed from 4.2.0.34 but I had the same issue. I'll try to build from master and will report back. Thx.

**Comment by istranic** [comment]
@adityak2920 Do you have any reliable instructions for installing form the github repo on Windows 10. I'm using the instructions below and am having some issues:

https://docs.opencv.org/master/d3/d52/tutorial_windows_install.html
